presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -849,4 +969,61 @@ public FunctionMap(FunctionMap map, Iterable<? extends SqlFunction> functions)
             return functions.get(name);
         }
     }
+
+    private class ApplicableFunction
+    {
+        private final Signature declaredSignature;
+        private final Signature boundSignature;
+
+        private ApplicableFunction(Signature declaredSignature, Signature boundSignature)
+        {
+            this.declaredSignature = declaredSignature;
+            this.boundSignature = boundSignature;
+        }
+
+        public Signature getDeclaredSignature()
+        {
+            return declaredSignature;
+        }
+
+        public Signature getBoundSignature()
+        {
+            return boundSignature;
+        }
+
+        /**
+         * One method is more specific than another if invocation handled by the first method could be passed on to the other one
+         */
+        public boolean isMoreSpecificThan(ApplicableFunction other)
```

Comment:
I think the abstraction would be a bit cleaner if this method was part of FunctionRegistry, since the TypeRegistry is what's needed to compute the answer.

This is, basically, a "comparator" of ApplicableFunction that requires state external to the objects.